Sources/ChatToys/Utils/WebContext.swift

```diff
@@ -194,3 +205,88 @@ extension Array {
     }
 }
 
+public class MarkdownProcessor {
+    private static var lock = NSLock()
+    private static var stringToKeyMap = [String: String]()
+    private static var uuidToURLMap = [String: String]()
+    public static var uuidLength = 5
+    
+    public static func updateStringToKeyMap(key: String, value: String) {
+        lock.lock()
+        defer { lock.unlock() }
+        stringToKeyMap[key] = value
+    }
+
+    public static func updateUuidToURLMap(key: String, value: String) {
+        lock.lock()
+        defer { lock.unlock() }
+        uuidToURLMap[key] = value
+    }
+    
+    public static func key(for value: String) -> String? {
+        return stringToKeyMap.first { $0.value == value }?.key
+    }
+
+    public static func url(for value: String) -> URL? {
+        guard let textContent = key(for: value)?.nilIfEmptyOrJustWhitespace,
+              var urlString = uuidToURLMap[value] else {
+            return nil
+        }
+        
+        let words = textContent.split(separator: " ")
+        
+        // TODO see if a more robust implementation would be warranted
+        if let range = urlString.range(of: "#:~:text=") {
+            let urlString = urlString[..<range.lowerBound]
+        }
+        
+        if words.count <= 8 {
+            urlString = "\(urlString)#:~:text=\( processWords(words.prefix(8)) )"
+        } else {
+            let start = processWords(words.prefix(4))
+            let end = processWords(words.suffix(4))
+            urlString = "\(urlString)#:~:text=\(start),\(end)"
+        }
+        
+        return URL(string: urlString)
+    }
+
+    private static func processWords(_ words: ArraySlice<Substring>) -> String {
+        return words.joined(separator: " ")
+            .addingPercentEncoding(withAllowedCharacters: .urlQueryAllowed) ?? ""
+            .replacingOccurrences(of: ",", with: "%2C")
+            .replacingOccurrences(of: "*", with: "")
+            .replacingOccurrences(of: "#", with: "")
+            .replacingOccurrences(of: "-", with: "")
+            .replacingOccurrences(of: "_", with: "")
+            .trimmingCharacters(in: .whitespaces)
+    }
+
+    public static func shortUUID(_ string: String) -> String {
+        if let existingKey = stringToKeyMap[string] {
+            return existingKey
+        }
+        var uuid: String
+        var collisions = 0
+        repeat {
+            uuid = String(UUID().uuidString.prefix(uuidLength))
+            if stringToKeyMap.values.contains(uuid) {
```

Comment:
i think reading this property also needs to be guarded by the lock!

An approach that would make this simpler:
Take the internal state (stringToKeyMap, uuidToURLMap, uuidLength) and wrap it in a struct. Then expose a method to modify this internal state, which handles locking and allows you to pass in a block that does the modification. Then you can make the actual var private and funnel all reads/writes thru this guarded `modify` function.

Other alternatives:
- Use an actor
- Treat this mapping of strings and keys as a value type that you pass _in_ when converting a string to markdown, then store it on the tab object